libs/utils/src/postgres_backend_async.rs

```diff
@@ -276,7 +276,10 @@ impl PostgresBackend {
         S: Future,
     {
         let ret = self.run_message_loop(handler, shutdown_watcher).await;
-        let _ = self.stream.shutdown();
+        let ignored = self.stream.shutdown().await;
+        if let Err(e) = ignored {
+            tracing::warn!("socket shutdown failed: {e:?}");
+        }
```

Comment:
This was not discussed on #2376: https://github.com/neondatabase/neon/pull/2376/files#r1131377891 (or the discussion has been replaced).

@hlinnaka do you agree with the change here?

Reply:
This must've been fixed with the rewrite from #3759.

Reply:
Unable to track it. @arssher do you remember? :)

Reply:
Found it in https://github.com/neondatabase/neon/blob/b80fe41af3e672f242ce4aab6699649f517d6e33/libs/postgres_backend/src/lib.rs#L366

Apologies on the noise unless this was a known fix.

Reply:
Yep. As (now) comment says, we don't accurately track state of the socket, so warning on shutdown failure would lead to a lot of noise. We can/should track it (and it would be good for other reasons), but I'm not sure how useful this particular shutdown warning anyway.

Reply:
I was thinking this might explain the #2963 flakyness if we fail to flush a tcp stream but I guess it shouldn't be hit here after all. Still, seems like an important thing to do.

> we don't accurately track state of the socket, so warning on shutdown failure would lead to a lot of noise

Possible... Has that ever been tried? It would though be just an std::io::Error, not like it carried a stacktrace.

Reply:
Yes, it looks like
```
2023-03-10T08:42:47.383607Z  INFO walreceiver_connection{id=773dbac0b63fb07ca3ccc4955148e050/5df9657b5afd9cbba25883e2470995c5 node_id=1}: caught up at LSN 0/16CC0D0
2023-03-10T08:42:48.088633Z  WARN socket shutdown failed: Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" }
2023-03-10T08:42:48.091778Z  WARN socket shutdown failed: Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" }
2023-03-10T08:42:48.091836Z  WARN socket shutdown failed: Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" }
```

Reply:
Oki thanks... A bit surprising error but perhaps something was still buffered. Missing-flush issues have just been historically for me quite interesting bugs, leading to a timeout which would hide some real issue.

I guess this is good to close. Thanks for the explanation.